src/lower/mod.rs

```diff
@@ -957,3 +981,49 @@ impl ir::TraitDatum {
         vec![wf]
     }
 }
+
+impl ir::AssociatedTyDatum {
+    fn to_program_clauses(&self) -> Vec<ir::ProgramClause> {
+        // For each associated type, we have a "normalization fallback" clause.
+        //
+        // Given:
+        //
+        //    trait SomeTrait {
+        //        type Assoc;
+        //    }
+        //
+        // we generate:
+        //
+        //    ?T: SomeTrait<Assoc = (SomeTrait::Assoc)<?T>> :-
```

Comment:
question -- currently, for each `type Foo = Bar` in an impl or the where-clause list, we generate *two* clauses:

    <T as Trait>::Foo ==> Bar

and 

    known { <T as Trait>::Foo ==> Bar }

Maybe we should just translate the impl declarations along with `T: Trait<Foo=Bar>` where-clauses into the "known" form to start, and then, for each declaration in the trait, have two instances of the "general rule":

    <?T as Trait>::Foo ==> Bar :- known { <?T as Trait>::Foo ==> Bar }
    <?T as Trait>::Foo ==> <?T as Trait>::Foo :- not { known { <?T as Trait>::Foo ==> Bar } }

Obviously, this is equivalent. Just seems a bit clearer to me somehow.